synapse/storage/databases/main/state.py

```diff
@@ -419,13 +419,15 @@ def _update_state_for_partial_state_event_txn(
         # anything that was rejected should have the same state as its
         # predecessor.
         if context.rejected:
-            assert context.state_group == context.state_group_before_event
```

Comment:
I wrote this in the git commit, but to reiterate, this was completely bogus. The value of `EventContext._state_group` on a rejected event is actually a dummy state group which *includes* the rejected event (since the state group is created before we know if the event will be rejected or not).

`EventContext.state_group` is a getter which has an assertion to check you aren't shooting yourself in the foot, but the right thing to do here is not to access it for rejected events.